### ISSUE TITLE
Bug: [prefer-regexp-exec] False positive when regex is an object property

### ISSUE THREAD

**Issue by RedGuy12** [typescript-eslint/typescript-eslint#8614]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.3.3&fileType=.ts&code=MYewdgzgLgBFCmAPWBeGByCIC28oAsBLMAc3QG4AoUSWEAIwCsAuGAbxgngEMAnYfKwBK8EgFFEABxgBfKgmQA6bNygCAFA0aKufAQEpyQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Y6RAM0Wls4DmiAB7F6wxGXRRe0APbRI4MAF8QKoA&tsconfig=&tokens=false

### Repro Code

```TypeScript
const text = 'something';
const obj: { search: RegExp };
text.match(obj.search);
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/prefer-regexp-exec": "error"
  }
}
```

### Expected Result

I expected no error to be reported since the regular expression's flags are unknown, and changing this to `String#match` may change runtime behavior.

### Actual Result

An error was reported
> Use the `RegExp#exec()` method instead. 3:6 - 3:11

### Additional Info

Related to https://github.com/typescript-eslint/typescript-eslint/issues/3477#issuecomment-922385173

**Comment by kirkwaiblinger** [comment]
related: https://github.com/typescript-eslint/typescript-eslint/issues/3477#issuecomment-922385173

**Comment by kirkwaiblinger** [comment]
Looks like the issue is that the rule uses whether an Identifier is passed to String.match as a heuristic for whether it may have the `/g` flag. 

See https://github.com/typescript-eslint/typescript-eslint/blob/fc08f969e255b55ea35091b297b358bc434a76eb/packages/eslint-plugin/src/rules/prefer-regexp-exec.ts#L75-L90

Without having thought too hard about it, I note that replacing the last line with `return true;` causes all the existing test cases to pass, as well as giving the expected behavior in your case

**Comment by Josh-Cena** [comment]
The heuristic does look weird to me. We should treat any non-analyzable expression (where the definition is "not regex literal or fresh constructor call", unless we have the ability to trace definitions) as potentially global.
